### ISSUE TITLE
Temp tables does not work with SELECT INTO statement. How to handle them with connection pooling?

### ISSUE THREAD

**Issue by YinDelphinus** [microsoft/mssql-jdbc#1547]
## Driver version
10.00.19041.01

## SQL Server version
Microsoft SQL Server 2019 (RTM) - 15.0.2000.5 (X64) 
	Sep 24 2019 13:48:23 
	Copyright (C) 2019 Microsoft Corporation
	Standard Edition (64-bit) on Windows 10 Pro 10.0 <X64> (Build 19041: ) (Hypervisor)

## Client Operating System
Windows 10 Pro

## JAVA/JVM version
openjdk version "15.0.1" 2020-10-20
OpenJDK Runtime Environment (build 15.0.1+9-18)
OpenJDK 64-Bit Server VM (build 15.0.1+9-18, mixed mode, sharing)

## Dependencies
mssql-jdbc-7.4.1.jre8.jar (current)
mssql-jdbc-9.2.1.jre11.jar (also tested, same result)
HikariCP-3.4.5.jar
spring-jdbc-5.2.8.RELEASE.jar

## Problem description
I want to use temp tables in my java application. Using hikari pool and spring-jdbc. I've seen this [issue](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/67b10d74-e94f-4b58-8ee3-59ff390db1cf/jdbc-and-temp-tables?forum=sqldataaccess) and this #1131, but for me # temp tables does not work with SELECT INTO statements. For some reason it works fine with ##. The general question is: how to do it properly, to be sure everything will be ok on prod?

By using @Transactional I can guarantee only that exact Connection will be kept alive and used during all jdbc calls in function. But I can't understand where sessions comes out? Where I can see their creation or closure? I don't want to get some heisenbug cause my session was closed and temp table deleted. Is there some special session-logic for SELECT INTO statements, or prepared statements?

Here is my test, currently we are using code like in selectIntoPrepLocal or selectIntoPrepGlobal, but with real tables to avoid heisenbugs:

`
    private JdbcTemplate jdbcTemplate;
    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;

    @Autowired
    private DataSource dataSource;

    @PostConstruct
    private void init() {
        this.jdbcTemplate = new JdbcTemplate(dataSource);
        this.namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(jdbcTemplate);
        QueryContextHolder.setCompanyNumber("1");
    }

    @Test
    @Transactional
    public void selectIntoPrepLocal() throws Exception {
        MapSqlParameterSource parameterSource = new MapSqlParameterSource("id", 1);
        namedParameterJdbcTemplate.update("SELECT ID INTO #TEMP1 FROM Employee WHERE ID = :id ", parameterSource);
        jdbcTemplate.execute("SELECT COUNT(*) FROM #TEMP1");
    }

    @Test
    @Transactional
    public void selectIntoPrepGlobal() throws Exception {
        MapSqlParameterSource parameterSource = new MapSqlParameterSource("id", 1);
        namedParameterJdbcTemplate.update("SELECT ID INTO ##TEMP1 FROM Employee WHERE ID = :id ", parameterSource);
        jdbcTemplate.execute("SELECT COUNT(*) FROM ##TEMP1");
    }

    @Test
    @Transactional
    public void selectIntoStmt() throws Exception {
        jdbcTemplate.update("SELECT ID INTO #TEMP1 FROM Employee WHERE ID = 1 ");
        jdbcTemplate.execute("SELECT COUNT(*) FROM #TEMP1");
    }

    @Test
    @Transactional
    public void createPrep(){
        jdbcTemplate.update("CREATE TABLE #TEMP1 ([ID] [bigint] NOT NULL)");
        MapSqlParameterSource parameterSource = new MapSqlParameterSource("id", 1);
        namedParameterJdbcTemplate.update("INSERT INTO #TEMP1 SELECT ID FROM Employee WHERE ID = :id ", parameterSource);
        jdbcTemplate.execute("SELECT COUNT(*) FROM #TEMP1");
    }

    @Test
    @Transactional
    public void createStmt(){
        jdbcTemplate.update("CREATE TABLE #TEMP1 ([ID] [bigint] NOT NULL)");
        jdbcTemplate.update("INSERT INTO #TEMP1 SELECT ID FROM Employee WHERE ID = 1 ");
        jdbcTemplate.execute("SELECT COUNT(*) FROM #TEMP1");
    }
`

All tests works fine, exept selectIntoPrepLocal, it throws an exception:

> StatementCallback; bad SQL grammar [SELECT COUNT(*) FROM #TEMP1]; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Invalid object name '#TEMP1'.
org.springframework.jdbc.BadSqlGrammarException: StatementCallback; bad SQL grammar [SELECT COUNT(*) FROM #TEMP1]; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: Invalid object name '#TEMP1'.
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:239)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:72)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1443)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:388)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:418)
	at com.upitec.erp.employeesrv.integrationtest.TempTablesTest.selectIntoPrepLocal(TempTablesTest.java:36) <internal calls>
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Invalid object name '#TEMP1'.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1624)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.doExecuteStatement(SQLServerStatement.java:868)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement$StmtExecCmd.doExecute(SQLServerStatement.java:768)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7194)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2979)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:248)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:223)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.execute(SQLServerStatement.java:744)
	at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:95)
	at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java)
	at org.springframework.jdbc.core.JdbcTemplate$1ExecuteStatementCallback.doInStatement(JdbcTemplate.java:409)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:376)
	... 90 more

**Comment by lilgreenbird** [comment]
hi @YinDelphinus 
We will investigate and get back to you.

**Comment by lilgreenbird** [comment]
hi @YinDelphinus 

I can't tell from the code you provided as it calls some unknown JdbcTemplate class but I think you might be using a different connection? A `#tempTable` is a local temporary table that is available only to the session that created it and dropped when closed. If you want the temp table to be visible and available to all your sessions you should use a global `##tempTable`

**Comment by YinDelphinus** [comment]
I'm afraid that the problem is not in the spring-jdbc. Even if will use pure ms-sql-jdbc lib, exact single connection that I will unwrap from hikari, and prepared statement without params, the temp table is invisible for next select. Sorry for I didn't start with that, here is new test:

`@SpringBootTest(classes = {DBContextConfig.class})
public class TempTablesMSTest {

    @Autowired
    private DataSource dataSource;

    @PostConstruct
    private void init() {
        QueryContextHolder.setCompanyNumber("1");
    }

    @Test
    public void selectIntoStmt() throws Exception {
        Connection connection = dataSource.getConnection().unwrap(SQLServerConnection.class);
        Statement statement = connection.createStatement();
        statement.executeUpdate("SELECT ID INTO [#TEMP_123] FROM Employee ORDER BY ID OFFSET 10 ROWS FETCH NEXT 10 ROWS ONLY");

        PreparedStatement statement1 = connection.prepareStatement("SELECT COUNT(*) FROM [#TEMP_123]");
        ResultSet resultSet = statement1.executeQuery();
        resultSet.next();
        System.out.println(resultSet.getInt(1));
    }

    @Test
    public void selectIntoPrp() throws Exception {
        Connection connection = dataSource.getConnection().unwrap(SQLServerConnection.class);
        PreparedStatement statement1 = connection
                .prepareStatement("SELECT ID INTO [#TEMP_123] FROM Employee ORDER BY ID OFFSET 10 ROWS FETCH NEXT 10 ROWS ONLY");
        statement1.executeUpdate();

        statement1 = connection.prepareStatement("SELECT COUNT(*) FROM [#TEMP_123]");
        ResultSet resultSet = statement1.executeQuery();
        resultSet.next();
        System.out.println(resultSet.getInt(1));
    }

    @Test
    public void selectStmt() throws Exception {
        Connection connection = dataSource.getConnection().unwrap(SQLServerConnection.class);;
        Statement statement = connection.createStatement();
        statement.executeUpdate("SELECT ID INTO [#TEMP_123] FROM Employee ORDER BY ID OFFSET 10 ROWS FETCH NEXT 10 ROWS ONLY");

        Statement statement1 = connection.createStatement();
        ResultSet resultSet = statement1.executeQuery("SELECT COUNT(*) FROM [#TEMP_123]");;
        resultSet.next();
        System.out.println(resultSet.getInt(1));
    }
}`

Here selectIntoPrp will fall with exception:

> com.microsoft.sqlserver.jdbc.SQLServerException: Invalid object name '#TEMP_123'.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1624)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:594)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:524)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7194)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2979)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:248)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:223)
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeQuery(SQLServerPreparedStatement.java:446)
	at com.upitec.erp.employeesrv.integrationtest.TempTablesMSTest.selectIntoPrp(TempTablesMSTest.java:48) <interanal calls>

The problem for me is that in my application I have to use connection pool for better performance. And like in all pools, connections has a max life time, so I can't just use global temp tables. If somewhere I have connection/session switch while using local temp table I will get an exception, but with global it can be a heisenbug where connection/session that created a temp table sometimes will just disappear before I finished my work with them, cause end of max life time or etc. So I can't just use gloabal temp tables, I need a guarantee that connection/session didn't switch. By using @Transactional I can be sure I'm using the same connection, but seems that something goes wrong anyway.

**Comment by YinDelphinus** [comment]
I think I found a [solution](https://itnext.io/net-parameterized-queries-issues-with-sql-server-temp-tables-f5ee4f304523). Here I can see that creating in a statement and than querying in a prepared will work fine. Can you approve that?

**Comment by lilgreenbird** [comment]
The issue as I mentioned in previous reply and also in #1131, is that you need to make sure the application is accessing the temporary table from the session that it was created. In your example where you are using 2 PreparedStatements that is 2 different sessions so a **local** temp table `#tempTable` will only be visible in the **local** session but not the other. If you want the temp table to be visible and available to all your sessions you should use a **global** `##tempTable` (with the 2 '#'). Fyi here is a [link](https://stackoverflow.com/questions/2920836/local-and-global-temporary-tables-in-sql-server) to stackoverflow that explains the difference between local and global temporary tables in SQL Server.  Or, as you pointed out, you could also use a statement as well to access the local temp table as that will be executed in the same session.

**Comment by YinDelphinus** [comment]
Ok, I understood. The last question - so my global temp table created in stored procedure will live until my connection is alive? No matter about lifetime of session of stored procedure?

**Comment by YinDelphinus** [comment]
Let me rephrase - how long the session of stored procedure that created temp table will live?

**Comment by YinDelphinus** [comment]
Global temp table ## will be alive until connection (not session) will be alive. Thanks!
